### ISSUE TITLE
How to modify the specific weights in the parallel models (13b)

### ISSUE THREAD

**Issue by cnlnpjhsy** [meta-llama/llama#1089]
I'm currently researching on the behaviors of FFN activations in llama-2 13b. I tried to collect the activation scores of FeedForward layer by storing the result of `F.silu(self.w1(x))`, and during the inference I disturb some columns of `w1.weight.T` to see their effects on generation (something like the code below). This method works well in llama-2-7b-chat, as the model only has one .pth file.
```
    original_w1: nn.Parameter = generator.model.layers[layer_idx].feed_forward.w1.weight.T      # transposed in the weight
    disrupted_w1 = original_w1.clone()
    disrupted_w1[:, disrupt_column_idx] += noise[:, disrupt_column_idx]
    generator.model.layers[layer_idx].feed_forward.w1.weight = nn.Parameter(disrupted_w1.T)
```

However, when I switched to llama-2-13b-chat, this method no longer works; I suppose that FFN parameters are stored in two checkpoint models. For an explicit example, in llama-2-13b-chat the parameter `w1` is defined as `self.w1 = ColumnParallelLinear(dim, hidden_dim, bias=False, gather_output=False, init_method=lambda x: x)`, with dim=5120 and hidden_dim=13824. When I directly access the `w1.weight.T` in the instantiated model, it only has half the column numbers 6912 because the parameters are distributed in different processes. I cannot get the other half activation scores using the old way :(

What I want to do is the same as 7b model: collect the activation scores of `F.silu(self.w1(x))` with full 13824 dimensions, and modify specific dimensions during the inference. I know this issue may be more associated with the torch usages, but still I'm hoping some ideas. Thanks! 🥲

**Comment by cnlnpjhsy** [comment]
So now I can collect the full activation scores by modifying the fairscale parameters:
```
        self.w1 = ColumnParallelLinear(
            dim, hidden_dim, bias=False, gather_output=True, init_method=lambda x: x
        )
        self.w2 = RowParallelLinear(
            hidden_dim, dim, bias=False, input_is_parallel=False, init_method=lambda x: x
        )
        self.w3 = ColumnParallelLinear(
            dim, hidden_dim, bias=False, gather_output=True, init_method=lambda x: x
        )
```
but I still cannot modify the specific column of weights; the `w1.weight` has still half the dimensions. Any body can help? 🥺

**Comment by cnlnpjhsy** [comment]
Finally I dig deeper into fairscale and write a subclass to solve the problem
```
class ColumnParallelLinearCustom(ColumnParallelLinear):
    def get_master_weight(self) -> torch.Tensor:
        from fairscale.nn.model_parallel.mappings import gather_from_model_parallel_region
        return gather_from_model_parallel_region(self.weight.data.transpose(0, 1).contiguous()).transpose_(0, 1)
    
    def update_master_weight(self, new_master_weight, stride=1):
        from fairscale.nn.model_parallel.initialize import get_model_parallel_rank, get_model_parallel_world_size
        from fairscale.nn.model_parallel.utils import divide_and_check_no_remainder
        self.master_weight = new_master_weight
        world_size = get_model_parallel_world_size()
        per_partition_per_stride_size = divide_and_check_no_remainder(self.output_size_per_partition, stride)
        weight_list = torch.split(new_master_weight, per_partition_per_stride_size, dim=0)
        rank = get_model_parallel_rank()
        my_weight_list = weight_list[rank::world_size]
        with torch.no_grad():
            torch.cat(my_weight_list, dim=0, out=self.weight)
```
... so this might actually be a fairscale issue. Closed!
